### PR TITLE
Feature/ldap gssapi sasl authentication

### LINKED ISSUE #1523: Support GSSAPI authentication for ldap_* module
##### SUMMARY

Add the attribute `sasl_class` with values `external` (default) or `gssapi` to `community.general/plugins/module_utils/ldap.py`
to select the SASL authentication algorithm for python-ldap when user name and password are not provide. Currently, the ansible modules default to EXTERNAL.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

ldap_attr
ldap_attrs
ldap_entry
ldap_passwd
ldap_search

##### ADDITIONAL INFORMATION

Adding the attribute `sasl_class` would allow to switch the default SASL authentication mechanism from `EXTERNAL` to `GSSAPI` which would help windows-based shops to transparently authenticate to Active Directory and use the LDAP modules to query and adjust their AD servers even from non-Windows servers. 

```yaml
- name: admin group must exist
  community.general.ldap_entry:
    dn: CN=my_admin_group,OU=Groups,OU=Global,DC=dir,DC=example,DC=com
    objectClass: group
    server_uri: "{{ ad.url }}"
    sasl_class: gssapi
    state: present
```

### PR DESCRIPTION
**Pull request by wimvanleuven**
##### SUMMARY

The LDAP modules support SASL passwordless authentications on systems that are properly setup. One such setup that python_ldap supports is Kerberos authentication to Windows domain. 

To allow parametrization of the SASL connection, this change adds the sasl_class parameter to LDAP modules, which defaults to 'external', but can also be set to 'gssapi'. 

Fixes #1523

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

plugins/modules/net_tools/ldap/ldap_attr.py
plugins/modules/net_tools/ldap/ldap_attrs.py
plugins/modules/net_tools/ldap/ldap_entry.py
plugins/modules/net_tools/ldap/ldap_passwd.py
plugins/modules/net_tools/ldap/ldap_search.py

##### ADDITIONAL INFORMATION

When using LDAP modules without userid and password for authentication, the implementation reverts to SASL passwordless authentication. For this to work, you system must be properly configured. In other environments, where Unix/Linux servers are integrated with Active Directory or Windows domain, using GSSAPI (Kerberos) for authentication makes more sense. 
The only requirement is to initialize the session with another class for interactive binding. 

Adding an additional parameters 'sasl_class' to 'plugins/module_utils/ldap.py' all LDAP modules allow this authentication system. 
If we default to 'external' we maintain backwards compatibility.

```paste below
- name: service account must exist and be member of admin group 
  community.general.ldap_attrs:
    dn: CN=sa_app_account,OU=Special Accounts,DC=dir,DC=example,DC=com
    attributes:
      memberOf: "CN=grp_app_admin_users,OU=Groups,DC=dir,DC=example,DC=com"
    state: exact
    server_uri: "{{ ad.url }}"
    sasl_class: gssapi 
  check_mode: yes
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -384,4 +446,4 @@ $RECYCLE.BIN/
 # Windows shortcuts
 *.lnk
 
-# End of https://www.gitignore.io/api/git,linux,pydev,python,windows,pycharm+all,jupyternotebook,vim,webstorm,emacs,dotenv
+# End of https://www.gitignore.io/api/git,linux,pydev,python,windows,pycharm+all,jupyternotebook,vim,webstorm,emacs,dotenv,vscode
```
Comment (Akasurde) [inline .gitignore:449]: Add `.gitignore` as part of separate PR.

File: plugins/doc_fragments/ldap.py [hunk 2]
```diff
@@ -44,4 +44,10 @@ class ModuleDocFragment(object):
       - This should only be used on sites using self-signed certificates.
     type: bool
     default: yes
+  sasl_class:
+    description:
+      - The class to use for SASL authentication
+      - possible choices: external, gssapi
+    type: str
+    default: external
```
Comment (Akasurde) [inline plugins/doc_fragments/ldap.py:52]:
```suggestion
    default: external
    version_added: "2.0.0"
```

File: plugins/doc_fragments/ldap.py [hunk 3]
```diff
@@ -44,4 +44,10 @@ class ModuleDocFragment(object):
       - This should only be used on sites using self-signed certificates.
     type: bool
     default: yes
+  sasl_class:
+    description:
+      - The class to use for SASL authentication
```
Comment (Akasurde) [inline plugins/doc_fragments/ldap.py:49]:
```suggestion
      - The class to use for SASL authentication.
```

File: plugins/doc_fragments/ldap.py [hunk 4]
```diff
@@ -44,4 +44,10 @@ class ModuleDocFragment(object):
...
+      - possible choices: external, gssapi
```
Comment (Akasurde) [inline plugins/doc_fragments/ldap.py:50]:
```suggestion
      - Possible values are C(external), C(gssapi).
```

File: plugins/module_utils/ldap.py [hunk 5]
```diff
@@ -71,7 +78,8 @@ def _connect_to_ldap(self):
             if self.bind_dn is not None:
                 connection.simple_bind_s(self.bind_dn, self.bind_pw)
             else:
-                connection.sasl_interactive_bind_s('', ldap.sasl.external())
+                klass = SASCL_CLASS.get(self.sasl_class, ldap.sasl.external)
```
Comment (Akasurde) [inline plugins/module_utils/ldap.py:81]: Would it possible to add integration tests for this change?

File: plugins/doc_fragments/ldap.py [hunk 6]
```diff
@@ -44,4 +44,10 @@ class ModuleDocFragment(object):
       - This should only be used on sites using self-signed certificates.
     type: bool
     default: yes
+  sasl_class:
```
Comment (Akasurde) [inline plugins/doc_fragments/ldap.py:47]: Add a change log entry for this change.

File: .gitignore (hunk 1 above)
Comment (wimvanleuven) [reply .gitignore:449]: I'll revert this

File: plugins/module_utils/ldap.py (hunk 5 above)
Comment (wimvanleuven) [reply plugins/module_utils/ldap.py:81]: I was thinking about that ... I did find a simple ldap_search integration test but not sure how to set this up for authentication schemes.

Have any pointers?

File: plugins/doc_fragments/ldap.py (hunk 2 above)
Comment (wimvanleuven) [reply plugins/doc_fragments/ldap.py:52]: done

File: plugins/doc_fragments/ldap.py (hunk 3 above)
Comment (wimvanleuven) [reply plugins/doc_fragments/ldap.py]: done

File: plugins/doc_fragments/ldap.py (hunk 6 above)
Comment (wimvanleuven) [reply plugins/doc_fragments/ldap.py:47]: done

File: changelogs/fragments/1595-ldap-gssapi-sasl-authentication.yml [hunk 7]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+- "ldap modules - add sasl_class parameter to support passwordless SASL authentication via GSSAPI (kerberos), next to external"
```
Comment (felixfontein) [inline changelogs/fragments/1595-ldap-gssapi-sasl-authentication.yml:2]:
```suggestion
- "ldap modules - add ``sasl_class`` parameter to support passwordless SASL authentication via GSSAPI (kerberos), next to external (https://github.com/ansible-collections/community.general/issues/1523)."
```

File: plugins/module_utils/ldap.py (hunk 5 above)
Comment (felixfontein) [reply plugins/module_utils/ldap.py:81]: You probably have to modify setup_openldap (https://github.com/ansible-collections/community.general/tree/main/tests/integration/targets/setup_openldap/) to allow both authentication forms. No idea how feasible that is though...

Comment (felixfontein) [reply plugins/module_utils/ldap.py:81]: (Afterwards you could use the ldap_search tests to use both methods.)

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@Akasurde I'm planing to merge this in a couple of days, if you don't object.

**Comment by felixfontein** [comment]
Restarting CI.

**Comment by felixfontein** [comment]
@wimvanleuven thanks for implementing this!
@Akasurde thanks for reviewing!

**Comment by wimvanleuven** [comment]
@felixfontein @Akasurde thanks for helping out and making this a smooth ride ... 
I did not find the time to check if I could add integration tests for it ... I'll try to complement this later ...

**Comment by felixfontein** [comment]
@wimvanleuven if you could do that, we would really appreciate that :)

**Pull request merged**